// altair-graphql-core/build/theme/defaults/dark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const theme_1 = require("../theme");
const theme = {
    colors: {
        bg: theme_1.foundations.colors.black,
        offBg: '#3f4349',
        font: theme_1.foundations.colors.white,
        offFont: theme_1.foundations.colors.lightGray,
        border: '#565c64',
        offBorder: '#565656',
        headerBg: '#3f4349',
    },
    shadow: {
        opacity: .3,
    },
    editor: {
        colors: {
            comment: theme_1.foundations.colors.darkGray,
            string: theme_1.foundations.colors.orange,
            number: theme_1.foundations.colors.orange,
            variable: theme_1.foundations.colors.white,
            attribute: theme_1.foundations.colors.green,
            keyword: theme_1.foundations.colors.blue,
            atom: theme_1.foundations.colors.white,
            property: theme_1.foundations.colors.blue,
            definition: theme_1.foundations.colors.orange,
            punctuation: theme_1.foundations.colors.blue,
            cursor: theme_1.foundations.colors.blue,
        }
    },
};
exports.default = theme;
