// @udecode/plate-line-height/dist/index.es.js
import { createPluginFactory, getPluginType, ELEMENT_DEFAULT, mapInjectPropsToPlugin, getPluginInjectProps, unsetNodes, setElements, isBlock } from '@udecode/plate-common';

const KEY_LINE_HEIGHT = 'lineHeight';
/**
 * Enables support for text alignment, useful to align your content
 * to left, right and center it.
 */

const createLineHeightPlugin = createPluginFactory({
  key: KEY_LINE_HEIGHT,
  inject: {
    props: {
      nodeKey: KEY_LINE_HEIGHT,
      defaultNodeValue: 1.5
    }
  },
  then: editor => ({
    inject: {
      props: {
        validTypes: [getPluginType(editor, ELEMENT_DEFAULT)]
      }
    },
    then: (_, plugin) => mapInjectPropsToPlugin(editor, plugin, {
      deserializeHtml: {
        getNode: (el, node) => {
          if (el.style.lineHeight) {
            node[plugin.key] = el.style.lineHeight;
          }
        }
      }
    })
  })
});

const setLineHeight = (editor, {
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = getPluginInjectProps(editor, KEY_LINE_HEIGHT);

  const match = n => isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);

  if (value === defaultNodeValue) {
    unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    setElements(editor, {
      [nodeKey]: value
    }, {
      match: match,
      ...setNodesOptions
    });
  }
};

export { KEY_LINE_HEIGHT, createLineHeightPlugin, setLineHeight };
