// @graphql-inspector/core/cjs/diff/changes/union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unionMemberAdded = exports.unionMemberRemoved = void 0;
const change_js_1 = require("./change.js");
function unionMemberRemoved(union, type) {
    return {
        criticality: {
            level: change_js_1.CriticalityLevel.Breaking,
            reason: 'Removing a union member from a union can cause existing queries that use this union member in a fragment spread to error.',
        },
        type: change_js_1.ChangeType.UnionMemberRemoved,
        message: `Member '${type.name}' was removed from Union type '${union.name}'`,
        path: union.name,
    };
}
exports.unionMemberRemoved = unionMemberRemoved;
function unionMemberAdded(union, type) {
    return {
        criticality: {
            level: change_js_1.CriticalityLevel.Dangerous,
            reason: 'Adding a possible type to Unions may break existing clients that were not programming defensively against a new possible type.',
        },
        type: change_js_1.ChangeType.UnionMemberAdded,
        message: `Member '${type.name}' was added to Union type '${union.name}'`,
        path: union.name,
    };
}
exports.unionMemberAdded = unionMemberAdded;
