// @udecode/plate-select/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var plateCore = require('@udecode/plate-core');

/**
 * Set a list of element types to select on backspace
 */
const withSelectOnBackspace = (editor, {
  options: {
    query
  }
}) => {
  const {
    deleteBackward
  } = editor;

  editor.deleteBackward = unit => {
    const {
      selection
    } = editor;

    if (unit === 'character' && plateCore.isCollapsed(selection)) {
      const pointBefore = plateCore.getPointBefore(editor, selection, {
        unit
      });

      if (pointBefore) {
        const [prevCell] = plateCore.getNodeEntries(editor, {
          match: node => plateCore.queryNode([node, pointBefore.path], query),
          at: pointBefore
        });

        if (!!prevCell && pointBefore) {
          // don't delete image, set selection there
          plateCore.select(editor, pointBefore);
        } else {
          deleteBackward(unit);
        }
      } else {
        deleteBackward(unit);
      }
    } else {
      deleteBackward(unit);
    }
  };

  return editor;
};

const KEY_SELECT_ON_BACKSPACE = 'selectOnBackspace';
/**
 * @see {@link withSelectOnBackspace}
 */

const createSelectOnBackspacePlugin = plateCore.createPluginFactory({
  key: KEY_SELECT_ON_BACKSPACE,
  withOverrides: withSelectOnBackspace
});

exports.KEY_SELECT_ON_BACKSPACE = KEY_SELECT_ON_BACKSPACE;
exports.createSelectOnBackspacePlugin = createSelectOnBackspacePlugin;
exports.withSelectOnBackspace = withSelectOnBackspace;
