// @udecode/plate-select/dist/index.es.js
import { isCollapsed, getPointBefore, getNodeEntries, queryNode, select, createPluginFactory } from '@udecode/plate-common';

/**
 * Set a list of element types to select on backspace
 */
const withSelectOnBackspace = (editor, {
  options: {
    query
  }
}) => {
  const {
    deleteBackward
  } = editor;

  editor.deleteBackward = unit => {
    const {
      selection
    } = editor;

    if (unit === 'character' && isCollapsed(selection)) {
      const pointBefore = getPointBefore(editor, selection, {
        unit
      });

      if (pointBefore) {
        const [prevCell] = getNodeEntries(editor, {
          match: node => queryNode([node, pointBefore.path], query),
          at: pointBefore
        });

        if (!!prevCell && pointBefore) {
          // don't delete image, set selection there
          select(editor, pointBefore);
        } else {
          deleteBackward(unit);
        }
      } else {
        deleteBackward(unit);
      }
    } else {
      deleteBackward(unit);
    }
  };

  return editor;
};

const KEY_SELECT_ON_BACKSPACE = 'selectOnBackspace';
/**
 * @see {@link withSelectOnBackspace}
 */

const createSelectOnBackspacePlugin = createPluginFactory({
  key: KEY_SELECT_ON_BACKSPACE,
  withOverrides: withSelectOnBackspace
});

export { KEY_SELECT_ON_BACKSPACE, createSelectOnBackspacePlugin, withSelectOnBackspace };
