// @udecode/plate-serializer-csv/dist/index.es.js
import { getPlugin, getPluginType, ELEMENT_DEFAULT, createPluginFactory } from '@udecode/plate-common';
import { ELEMENT_TABLE, ELEMENT_TH, ELEMENT_TR, ELEMENT_TD } from '@udecode/plate-table';
import { parse } from 'papaparse';

const isValidCsv = (data, errors, errorTolerance) => {
  if (errorTolerance < 0) errorTolerance = 0;
  return !(!data || data.length < 2 || data[0].length < 2 || data[1].length < 2 || errors.length && errors.length > errorTolerance * data.length);
};

const deserializeCsv = (editor, {
  data,
  ...parseOptions
}) => {
  const {
    options: {
      errorTolerance,
      parseOptions: pluginParseOptions
    }
  } = getPlugin(editor, KEY_DESERIALIZE_CSV); // Verify it's a csv string

  const testCsv = parse(data, {
    preview: 2
  });

  if (testCsv.errors.length === 0) {
    const csv = parse(data, { ...pluginParseOptions,
      ...parseOptions
    });
    if (!isValidCsv(csv.data, csv.errors, errorTolerance)) return;
    const paragraph = getPluginType(editor, ELEMENT_DEFAULT);
    const table = getPluginType(editor, ELEMENT_TABLE);
    const th = getPluginType(editor, ELEMENT_TH);
    const tr = getPluginType(editor, ELEMENT_TR);
    const td = getPluginType(editor, ELEMENT_TD);
    const ast = {
      type: table,
      children: []
    };

    if (csv.meta.fields) {
      // csv file has headers, data structure is an array of objects keyed on the heading title
      ast.children.push({
        type: tr,
        children: csv.meta.fields.map(field => ({
          type: th,
          children: [{
            type: paragraph,
            children: [{
              text: field
            }]
          }]
        }))
      });

      for (const row of csv.data) {
        ast.children.push({
          type: tr,
          children: csv.meta.fields.map(field => ({
            type: td,
            children: [{
              type: paragraph,
              children: [{
                text: row[field] || ''
              }]
            }]
          }))
        });
      }
    } else {
      // csv is an array of arrays
      for (const row of csv.data) {
        ast.children.push({
          type: tr,
          children: []
        });

        for (const cell of row) {
          ast.children[ast.children.length - 1].children.push({
            type: td,
            children: [{
              type: paragraph,
              children: [{
                text: cell
              }]
            }]
          });
        }
      }
    }

    return [{
      type: paragraph,
      children: [{
        text: ''
      }]
    }, ast, {
      type: paragraph,
      children: [{
        text: ''
      }]
    }];
  }
};

const KEY_DESERIALIZE_CSV = 'deserializeCsv';
/**
 * Enables support for deserializing content
 * from CSV format to Slate format.
 */

const createDeserializeCsvPlugin = createPluginFactory({
  key: KEY_DESERIALIZE_CSV,
  options: {
    errorTolerance: 0.25,
    parseOptions: {
      header: true
    }
  },
  then: editor => ({
    editor: {
      insertData: {
        format: 'text/plain',
        getFragment: ({
          data
        }) => deserializeCsv(editor, {
          data
        })
      }
    }
  })
});

export { KEY_DESERIALIZE_CSV, createDeserializeCsvPlugin, deserializeCsv };
