// altair-graphql-core/build/plugin/plugin.interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPlugin = exports.PluginType = exports.PluginSource = void 0;
/**
 * Defines the repository of the plugin.
 * Used to know where to get the plugin from.
 */
var PluginSource;
(function (PluginSource) {
    PluginSource["NPM"] = "npm";
    PluginSource["GITHUB"] = "github";
    PluginSource["URL"] = "url";
})(PluginSource = exports.PluginSource || (exports.PluginSource = {}));
/**
 * Specifies the type of the plugin.
 * Determines how the plugin would interact with Altair.
 */
var PluginType;
(function (PluginType) {
    PluginType["HEADER"] = "header";
    PluginType["SIDEBAR"] = "sidebar";
    PluginType["ACTION_BUTTON"] = "action_button";
})(PluginType = exports.PluginType || (exports.PluginType = {}));
exports.createPlugin = (name, manifest) => {
    return {
        name,
        manifest,
        type: manifest.type,
        display_name: manifest.display_name || name,
        plugin_class: manifest.plugin_class,
        capabilities: Array.from(new Set([...(manifest.capabilities || []), ...['query:read', 'query:write']])),
    };
};
