// @graphql-inspector/core/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateTokenCount = exports.countDepth = exports.countDirectives = exports.calculateOperationComplexity = exports.countAliases = exports.validate = exports.getTypePrefix = exports.similar = exports.CriticalityLevel = exports.ChangeType = exports.DiffRule = exports.diff = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./coverage/index.js"), exports);
var index_js_1 = require("./diff/index.js");
Object.defineProperty(exports, "diff", { enumerable: true, get: function () { return index_js_1.diff; } });
Object.defineProperty(exports, "DiffRule", { enumerable: true, get: function () { return index_js_1.DiffRule; } });
var change_js_1 = require("./diff/changes/change.js");
Object.defineProperty(exports, "ChangeType", { enumerable: true, get: function () { return change_js_1.ChangeType; } });
Object.defineProperty(exports, "CriticalityLevel", { enumerable: true, get: function () { return change_js_1.CriticalityLevel; } });
var index_js_2 = require("./similar/index.js");
Object.defineProperty(exports, "similar", { enumerable: true, get: function () { return index_js_2.similar; } });
var graphql_js_1 = require("./utils/graphql.js");
Object.defineProperty(exports, "getTypePrefix", { enumerable: true, get: function () { return graphql_js_1.getTypePrefix; } });
var index_js_3 = require("./validate/index.js");
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return index_js_3.validate; } });
var alias_count_js_1 = require("./validate/alias-count.js");
Object.defineProperty(exports, "countAliases", { enumerable: true, get: function () { return alias_count_js_1.countAliases; } });
var complexity_js_1 = require("./validate/complexity.js");
Object.defineProperty(exports, "calculateOperationComplexity", { enumerable: true, get: function () { return complexity_js_1.calculateOperationComplexity; } });
var directive_count_js_1 = require("./validate/directive-count.js");
Object.defineProperty(exports, "countDirectives", { enumerable: true, get: function () { return directive_count_js_1.countDirectives; } });
var query_depth_js_1 = require("./validate/query-depth.js");
Object.defineProperty(exports, "countDepth", { enumerable: true, get: function () { return query_depth_js_1.countDepth; } });
var token_count_js_1 = require("./validate/token-count.js");
Object.defineProperty(exports, "calculateTokenCount", { enumerable: true, get: function () { return token_count_js_1.calculateTokenCount; } });
