// @udecode/plate-trailing-block/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var plateCore = require('@udecode/plate-core');
var slate = require('slate');

/**
 * Add a trailing block when the last node type is not `type` and when the editor has .
 */
const withTrailingBlock = (editor, {
  type,
  options: {
    level,
    ...query
  }
}) => {
  const {
    normalizeNode
  } = editor;

  editor.normalizeNode = ([currentNode, currentPath]) => {
    if (!currentPath.length) {
      const lastChild = plateCore.getLastNodeByLevel(editor, level);
      const lastChildNode = lastChild === null || lastChild === void 0 ? void 0 : lastChild[0];

      if (!lastChildNode || lastChildNode.type !== type && plateCore.queryNode(lastChild, query)) {
        const at = lastChild ? slate.Path.next(lastChild[1]) : [0];
        plateCore.insertElements(editor, {
          type: type,
          children: [{
            text: ''
          }]
        }, {
          at
        });
        return;
      }
    }

    return normalizeNode([currentNode, currentPath]);
  };

  return editor;
};

const KEY_TRAILING_BLOCK = 'trailingBlock';
/**
 * @see {@link withTrailingBlock}
 */

const createTrailingBlockPlugin = plateCore.createPluginFactory({
  key: KEY_TRAILING_BLOCK,
  withOverrides: withTrailingBlock,
  options: {
    level: 0
  },
  then: editor => ({
    type: plateCore.getPluginType(editor, plateCore.ELEMENT_DEFAULT)
  })
});

exports.KEY_TRAILING_BLOCK = KEY_TRAILING_BLOCK;
exports.createTrailingBlockPlugin = createTrailingBlockPlugin;
exports.withTrailingBlock = withTrailingBlock;
