// tinacms/dist/rich-text/prism.js
(function(global, factory) {
  typeof exports === "object" && typeof module !== "undefined" ? factory(exports, require("react"), require("prism-react-renderer")) : typeof define === "function" && define.amd ? define(["exports", "react", "prism-react-renderer"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.tinacms = {}, global.NOOP, global.NOOP));
})(this, function(exports2, React, Highlight) {
  "use strict";
  function _interopDefaultLegacy(e) {
    return e && typeof e === "object" && "default" in e ? e : { "default": e };
  }
  var React__default = /* @__PURE__ */ _interopDefaultLegacy(React);
  var Highlight__default = /* @__PURE__ */ _interopDefaultLegacy(Highlight);
  var theme = {
    plain: {
      color: "#393A34",
      backgroundColor: "#f6f8fa"
    },
    styles: [{
      types: ["comment", "prolog", "doctype", "cdata"],
      style: {
        color: "#999988",
        fontStyle: "italic"
      }
    }, {
      types: ["namespace"],
      style: {
        opacity: 0.7
      }
    }, {
      types: ["string", "attr-value"],
      style: {
        color: "#e3116c"
      }
    }, {
      types: ["punctuation", "operator"],
      style: {
        color: "#393A34"
      }
    }, {
      types: ["entity", "url", "symbol", "number", "boolean", "variable", "constant", "property", "regex", "inserted"],
      style: {
        color: "#36acaa"
      }
    }, {
      types: ["atrule", "keyword", "attr-name", "selector"],
      style: {
        color: "#00a4db"
      }
    }, {
      types: ["function", "deleted", "tag"],
      style: {
        color: "#d73a49"
      }
    }, {
      types: ["function-variable"],
      style: {
        color: "#6f42c1"
      }
    }, {
      types: ["tag", "selector", "keyword"],
      style: {
        color: "#00009f"
      }
    }]
  };
  var theme$1 = theme;
  const Prism = (props) => {
    return /* @__PURE__ */ React__default["default"].createElement(Highlight__default["default"], {
      ...Highlight.defaultProps,
      theme: theme$1,
      code: props.value,
      language: props.lang
    }, ({ className, style, tokens, getLineProps, getTokenProps }) => /* @__PURE__ */ React__default["default"].createElement("pre", {
      className,
      style
    }, tokens.map((line, i) => /* @__PURE__ */ React__default["default"].createElement("div", {
      ...getLineProps({ line, key: i })
    }, line.map((token, key) => /* @__PURE__ */ React__default["default"].createElement("span", {
      ...getTokenProps({ token, key })
    }))))));
  };
  exports2.Prism = Prism;
  Object.defineProperties(exports2, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
});
