// @accessible/drawer/dist/main/index.js
"use strict";

exports.__esModule = true;
exports.useDrawer = useDrawer;
exports.Drawer = Drawer;
exports.useA11yTarget = useA11yTarget;
exports.Target = Target;
exports.useA11yCloseButton = exports.useA11yTrigger = exports.CloseButton = exports.Trigger = void 0;

var React = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _disclosure = /*#__PURE__*/require("@accessible/disclosure");

exports.Trigger = _disclosure.Trigger;
exports.CloseButton = _disclosure.CloseButton;
exports.useA11yTrigger = _disclosure.useA11yTrigger;
exports.useA11yCloseButton = _disclosure.useA11yCloseButton;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const __reactCreateElement__ = React.createElement;

/**
 * This hook provides the current value of the drawer's context object
 */
function useDrawer() {
  return (0, _disclosure.useDisclosure)();
}
/**
 * This component creates the context for your drawer target and trigger
 * and contains some configuration options.
 */


function Drawer(props) {
  return /*#__PURE__*/__reactCreateElement__(_disclosure.Disclosure, { ...props
  });
}
/**
 * A React hook for creating a headless drawer target to [WAI-ARIA authoring practices](https://www.w3.org/TR/wai-aria-practices/examples/dialog-drawer/dialog.html).
 *
 * @param target A React ref or HTML element
 * @param options Configuration options
 */


function useA11yTarget(target, options = {}) {
  const disclosureProps = (0, _disclosure.useA11yTarget)(target, Object.assign({}, options, {
    openStyle: Object.assign(defaultOpenStyles, options.openStyle)
  }));
  return Object.assign(disclosureProps, {
    style: Object.assign(disclosureProps.style, defaultClosedStyles[options.placement || 'left'])
  });
}
/**
 * This component wraps any React element and turns it into a
 * drawer target.
 */


function Target(props) {
  const childProps = props.children.props;
  return __reactCreateElement__(_disclosure.Target, Object.assign({}, props, {
    openStyle: Object.assign({}, defaultOpenStyles, props.openStyle)
  }), /*#__PURE__*/React.cloneElement(props.children, {
    style: Object.assign({}, defaultClosedStyles[props.placement || 'left'], childProps.style)
  }));
}

const defaultClosedStyles = {
  top: {
    position: 'fixed',
    top: 0,
    right: 0,
    bottom: 'auto',
    left: 0,
    transform: 'translate3d(0, -100%, 0)'
  },
  right: {
    position: 'fixed',
    top: 0,
    right: 0,
    bottom: 0,
    left: 'auto',
    transform: 'translate3d(100%, 0, 0)'
  },
  bottom: {
    position: 'fixed',
    top: 'auto',
    right: 0,
    bottom: 0,
    left: 0,
    transform: 'translate3d(0, 100%, 0)'
  },
  left: {
    position: 'fixed',
    top: 0,
    right: 'auto',
    bottom: 0,
    left: 0,
    transform: 'translate3d(-100%, 0, 0)'
  }
};
const defaultOpenStyles = {
  transform: 'translate3d(0, 0, 0)'
};

/* istanbul ignore next */
if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'production') {
  Drawer.displayName = 'Drawer';
  Target.displayName = 'Target';
}
